File: 0056-merge-intervals/0056-merge-intervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> res;
        sort(intervals.begin(), intervals.end());
        int start=intervals[0][0], end = intervals[0][1], len=intervals.size();

        for(int i=1; i<len ; i++){
            if(end >= intervals[i][0])
            {
                end = max(intervals[i][1],end);  // taking max bcz [[1,4],[2,3]] res: [1,4]
            }
            else{
                res.push_back({start,end});
                start = intervals[i][0];
                end = intervals[i][1];
            }


        }

        res.push_back({start,end});

        return res;
    }
};

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
  public int[][] merge(int[][] intervals) {
    // sort the 2D array
    Arrays.sort(intervals, (a,b) -> Integer.compare(a[0],b[0]));

    List<int[]> res = new ArrayList<>();
    int[] prev = intervals[0];

    for (int i = 1; i < intervals.length; i++) {
      if (intervals[i][0] <= prev[1]){ // start(0-index) of next interval is less than prev end
        prev[1] = Math.max(intervals[i][1], prev[1]); // {[1,4], [2,3]}    
      } else {
        res.add(prev);
        prev = intervals[i]; // storing current interval in prev for next iteration
      }
    }

    res.add(prev);

    return res.toArray(new int[res.size()][]); 
  }
}